crates/lint/src/linter/mod.rs

```diff
@@ -174,90 +178,98 @@ impl<'s, 'c> LintContext<'s, 'c> {
+// An emittable diagnostic suggestion.
+//
+// Depending on its [`SuggestionKind`] will be emitted as a simple note (examples), or a fix
+// suggestion.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Suggestion {
+    /// An optional description displayed above the code block.
+    desc: Option<&'static str>,
+    /// The actual suggestion.
+    content: String,
+    /// The suggestion type and its specific data.
+    kind: SuggestionKind,
+}
 
-        match self {
-            Self::Diff { span, add, trim_code: trim, .. } => {
-                // Get the original code from the span if provided and normalize its indentation
-                if let Some(span) = span
-                    && let Some(rmv) = ctx.span_to_snippet(span)
-                {
-                    let ind = ctx.get_span_indentation(span);
-                    let diag_msg = |line: &str, prefix: &str, style: Style| {
-                        let content = if trim { Self::trim_start_limited(line, ind) } else { line };
-                        (DiagMsg::from(format!("{prefix}{content}\n")), style)
-                    };
-                    output.extend(rmv.lines().map(|line| diag_msg(line, "- ", Style::Removal)));
-                    output.extend(add.lines().map(|line| diag_msg(line, "+ ", Style::Addition)));
-                } else {
-                    // Should never happen, but fall back to `Self::Block` behavior.
-                    output.extend(
-                        add.lines()
-                            .map(|line| (DiagMsg::from(format!("{line}\n")), Style::NoStyle)),
-                    );
-                }
-            }
-            Self::Block { code, .. } => {
-                output.extend(
-                    code.lines().map(|line| (DiagMsg::from(format!("{line}\n")), Style::NoStyle)),
-                );
-            }
+impl Suggestion {
+    /// Creates a new [`SuggestionKind::Example`] suggestion.
+    pub fn example(content: String) -> Self {
+        Self { desc: None, content, kind: SuggestionKind::Example }
+    }
+
+    /// Creates a new [`SuggestionKind::Fix`] suggestion.
+    ///
+    /// When possible, will attempt to inline the suggestion.
+    pub fn fix(content: String, applicability: Applicability) -> Self {
+        Self {
+            desc: None,
+            content,
+            kind: SuggestionKind::Fix {
+                span: None,
+                applicability,
+                style: SuggestionStyle::ShowCode,
+            },
         }
-        output.push((DiagMsg::from("\n"), Style::NoStyle));
-        output
     }
 
-    pub fn desc(&self) -> Option<&'static str> {
-        match self {
-            Self::Diff { desc, .. } => *desc,
-            Self::Block { desc, .. } => *desc,
+    /// Sets the description for the suggestion.
+    pub fn with_desc(mut self, desc: &'static str) -> Self {
+        self.desc = Some(desc);
+        self
+    }
+
+    /// Sets the span for a `[SuggestionKind::Fix]` suggestion.
```

Comment:
```suggestion
    /// Sets the span for a [`SuggestionKind::Fix`] suggestion.
```
should really be a clippy lint